File: ios/QingKeSchedule/Features/SemesterDraft.swift
import Foundation

struct PeriodDraft: Identifiable, Equatable {
    let id: UUID
    var number: Int
    var startTime: Date
    var endTime: Date

    init(
        id: UUID = UUID(),
        number: Int,
        startTime: Date,
        endTime: Date
    ) {
        self.id = id
        self.number = number
        self.startTime = startTime
        self.endTime = endTime
    }
}

struct SemesterDraft {
    var id: String
    var name: String
    var startDate: Date
    var totalWeeks: Int
    var periods: [PeriodDraft]

    init(
        semester: SemesterDTO? = nil,
        now: Date = Date(),
        calendar: Calendar = ScheduleRules.gregorianCalendar()
    ) {
        if let semester {
            id = semester.id
            name = semester.name
            startDate = ScheduleRules.localDate(
                from: semester.startDate,
                calendar: calendar
            ) ?? now
            totalWeeks = semester.totalWeeks
            periods = semester.periods.map {
                PeriodDraft(
                    number: $0.number,
                    startTime: Self.timeDate(from: $0.startTime, calendar: calendar),
                    endTime: Self.timeDate(from: $0.endTime, calendar: calendar)
                )
            }
        } else {
            let year = calendar.component(.year, from: now)
            let season = calendar.component(.month, from: now) >= 7 ? "秋季" : "春季"
            id = UUID().uuidString
            name = "\(year) \(season)学期"
            startDate = calendar.startOfDay(for: now)
            totalWeeks = 18
            periods = Self.defaultPeriodValues.enumerated().map { index, times in
                PeriodDraft(
                    number: index + 1,
                    startTime: Self.timeDate(from: times.0, calendar: calendar),
                    endTime: Self.timeDate(from: times.1, calendar: calendar)
                )
            }
        }
    }

    func semester(calendar: Calendar = ScheduleRules.gregorianCalendar()) -> SemesterDTO {
        SemesterDTO(
            id: id,
            name: name.trimmingCharacters(in: .whitespacesAndNewlines),
            startDate: Self.dateString(from: startDate, calendar: calendar),
            totalWeeks: totalWeeks,
            periods: periods.enumerated().map { index, period in
                PeriodDTO(
                    number: index + 1,
                    startTime: Self.timeString(from: period.startTime, calendar: calendar),
                    endTime: Self.timeString(from: period.endTime, calendar: calendar)
                )
            }
        )
    }

    func validationIssues(
        calendar: Calendar = ScheduleRules.gregorianCalendar()
    ) -> [ScheduleValidationIssue] {
        ScheduleValidator.validate(
            ScheduleDataDTO(
                semester: semester(calendar: calendar),
                courses: [],
                updatedAt: "1970-01-01T00:00:00.000Z"
            ),
            calendar: calendar
        )
    }

    mutating func addPeriod(calendar: Calendar = ScheduleRules.gregorianCalendar()) {
        let previousEnd = periods.last?.endTime
            ?? Self.timeDate(from: "07:50", calendar: calendar)
        let start = calendar.date(byAdding: .minute, value: 10, to: previousEnd) ?? previousEnd
        let end = calendar.date(byAdding: .minute, value: 45, to: start) ?? start
        periods.append(
            PeriodDraft(
                number: periods.count + 1,
                startTime: start,
                endTime: end
            )
        )
    }

    mutating func removePeriod(id: UUID) {
        guard periods.count > 1 else { return }
        periods.removeAll { $0.id == id }
        renumberPeriods()
    }

    mutating func renumberPeriods() {
        for index in periods.indices {
            periods[index].number = index + 1
        }
    }

    private static let defaultPeriodValues: [(String, String)] = [
        ("08:00", "08:45"),
        ("08:55", "09:40"),
        ("10:00", "10:45"),
        ("10:55", "11:40"),
        ("14:00", "14:45"),
        ("14:55", "15:40"),
        ("16:00", "16:45"),
        ("16:55", "17:40"),
        ("19:00", "19:45"),
        ("19:55", "20:40"),
    ]

    private static func dateString(from date: Date, calendar: Calendar) -> String {
        let components = calendar.dateComponents([.year, .month, .day], from: date)
        return String(
            format: "%04d-%02d-%02d",
            components.year ?? 0,
            components.month ?? 0,
            components.day ?? 0
        )
    }

    private static func timeString(from date: Date, calendar: Calendar) -> String {
        let components = calendar.dateComponents([.hour, .minute], from: date)
        return String(format: "%02d:%02d", components.hour ?? 0, components.minute ?? 0)
    }

    private static func timeDate(from value: String, calendar: Calendar) -> Date {
        let minutes = ScheduleRules.minutes(from: value) ?? 0
        var components = DateComponents()
        components.calendar = calendar
        components.timeZone = calendar.timeZone
        components.year = 2001
        components.month = 1
        components.day = 1
        components.hour = minutes / 60
        components.minute = minutes % 60
        return calendar.date(from: components) ?? Date(timeIntervalSinceReferenceDate: 0)
    }
}

File: Android/app/src/main/java/com/qingke/schedule/draft/SemesterDraft.kt
package com.qingke.schedule.draft

import com.qingke.schedule.domain.Course
import com.qingke.schedule.domain.Period
import com.qingke.schedule.domain.PeriodIdentity
import com.qingke.schedule.domain.ScheduleData
import com.qingke.schedule.domain.ScheduleValidator
import com.qingke.schedule.domain.Semester
import com.qingke.schedule.domain.ValidationIssue
import java.time.LocalDate
import java.time.LocalTime
import java.time.format.DateTimeFormatter
import java.util.UUID

data class PeriodDraft(
    val id: String,
    var number: Int,
    var startTime: LocalTime,
    var endTime: LocalTime,
    /** Number this period carried in the persisted semester; null for periods added inside the draft. */
    var sourceNumber: Int? = null,
)

class SemesterDraft private constructor(
    var id: String,
    var name: String,
    var startDate: LocalDate,
    var totalWeeks: Int,
    val periods: MutableList<PeriodDraft>,
    private val idFactory: () -> String,
) {
    fun semester(): Semester = Semester(
        id, name.trim(), startDate.format(DATE), totalWeeks,
        periods.map { Period(it.number, it.startTime.format(TIME), it.endTime.format(TIME)) },
    )

    /**
     * Non-continuable input checks that only exist because already saved courses are compared against the
     * draft: shortening the semester below a saved course range cannot be cascaded, so it must be fixed in
     * the course editor first. Period deletions no longer block here; P3-06-R7 turns them into a
     * [com.qingke.schedule.domain.SemesterCascadePlan] instead.
     */
    fun courseRangeIssues(previous: Semester?, courses: List<Course>): List<ValidationIssue> {
        if (previous == null || courses.isEmpty()) return emptyList()
        return if (courses.any { course -> course.schedules.any { it.endWeek > totalWeeks } }) {
            listOf(
                ValidationIssue(
                    "semester.totalWeeks",
                    "缩短总周数会让已有课程超出学期范围，请先在课程编辑中调整相关课程的周次。",
                ),
            )
        } else {
            emptyList()
        }
    }

    /** Identity of every visible row for [com.qingke.schedule.domain.SemesterCascadePlanner]. */
    fun periodIdentities(): List<PeriodIdentity> = periods.map { PeriodIdentity(it.sourceNumber, it.number) }

    /**
     * Re-points the identity baseline at a semester that was just written. [persistedNumbers] must
     * be the numbers captured when the write was requested, keyed by draft period id, so that edits
     * made while the write was in flight are not reported as persisted. Periods missing from that
     * map were created after the request and keep no persisted identity. Only call this after a
     * successful write.
     */
    fun markPersisted(persistedNumbers: Map<String, Int>) {
        periods.forEach { period -> period.sourceNumber = persistedNumbers[period.id] }
    }

    fun validationIssues(): List<ValidationIssue> = ScheduleValidator.validate(
        ScheduleData(1, semester(), emptyList(), "1970-01-01T00:00:00Z"),
    )

    fun addPeriod() {
        val start = (periods.lastOrNull()?.endTime ?: LocalTime.of(7, 50)).plusMinutes(10)
        val number = (periods.maxOfOrNull { it.number } ?: 0) + 1
        periods += PeriodDraft(idFactory(), number, start, start.plusMinutes(45))
    }

    fun removePeriod(id: String) {
        if (periods.size > 1) {
            periods.removeAll { it.id == id }
            periods.forEachIndexed { index, period -> period.number = index + 1 }
        }
    }

    companion object {
        fun create(now: LocalDate, idFactory: () -> String = { UUID.randomUUID().toString() }): SemesterDraft {
            val season = if (now.monthValue < 7) "春季" else "秋季"
            return SemesterDraft(idFactory(), "${now.year} ${season}学期", now, 18,
                DEFAULT_TIMES.mapIndexed { index, pair -> PeriodDraft(idFactory(), index + 1, pair.first, pair.second) }.toMutableList(), idFactory)
        }

        fun edit(semester: Semester, idFactory: () -> String = { UUID.randomUUID().toString() }): SemesterDraft = SemesterDraft(
            semester.id, semester.name, LocalDate.parse(semester.startDate), semester.totalWeeks,
            semester.periods.map {
                PeriodDraft(
                    idFactory(), it.number, LocalTime.parse(it.startTime), LocalTime.parse(it.endTime),
                    sourceNumber = it.number,
                )
            }.toMutableList(), idFactory,
        )

        private val DEFAULT_TIMES = listOf(
            "08:00" to "08:45", "08:55" to "09:40", "10:00" to "10:45", "10:55" to "11:40",
            "14:00" to "14:45", "14:55" to "15:40", "16:00" to "16:45", "16:55" to "17:40",
            "19:00" to "19:45", "19:55" to "20:40",
        ).map { LocalTime.parse(it.first) to LocalTime.parse(it.second) }
        private val DATE = DateTimeFormatter.ISO_LOCAL_DATE
        private val TIME = DateTimeFormatter.ofPattern("HH:mm")
    }
}
